tfsdk/serve.go

```diff
@@ -216,15 +216,73 @@ func (s *server) GetProviderSchema(ctx context.Context, _ *tfprotov6.GetProvider
 }
 
 func (s *server) ValidateProviderConfig(ctx context.Context, req *tfprotov6.ValidateProviderConfigRequest) (*tfprotov6.ValidateProviderConfigResponse, error) {
-	// uncomment when we implement this function
-	// ctx = s.registerContext(ctx)
-
-	// We don't actually do anything as part of this. In theory, we could
-	// validate the configuration for the provider block? Need to check in
-	// again with the core team about the goal of this RPC.
-	return &tfprotov6.ValidateProviderConfigResponse{
+	ctx = s.registerContext(ctx)
+	resp := &tfprotov6.ValidateProviderConfigResponse{
+		// This RPC allows a modified configuration to be returned. This was
+		// previously used to allow a "required" provider attribute (as defined
+		// by a schema) to still be "optional" with a default value, typically
+		// through an environment variable. Other tooling based on the provider
+		// schema information could not determine this implementation detail.
+		// To ensure accuracy going forward, this implementation is opinionated
+		// towards accurate provider schema definitions and optional values
+		// can be filled in or return errors during ConfigureProvider().
 		PreparedConfig: req.Config,
-	}, nil
+	}
+
+	schema, diags := s.p.GetSchema(ctx)
+
+	if diags != nil {
+		resp.Diagnostics = append(resp.Diagnostics, diags...)
+
+		if diagsHasErrors(resp.Diagnostics) {
+			return resp, nil
+		}
+	}
+
+	config, err := req.Config.Unmarshal(schema.TerraformType(ctx))
+
+	if err != nil {
+		resp.Diagnostics = append(resp.Diagnostics, &tfprotov6.Diagnostic{
+			Severity: tfprotov6.DiagnosticSeverityError,
+			Summary:  "Error parsing config",
+			Detail:   "The provider had a problem parsing the config. Report this to the provider developer:\n\n" + err.Error(),
+		})
+
+		return resp, nil
+	}
+
+	vpcReq := ValidateProviderConfigRequest{
+		Config: Config{
+			Raw:    config,
+			Schema: schema,
+		},
+	}
+
+	if provider, ok := s.p.(ProviderWithConfigValidators); ok {
+		for _, configValidator := range provider.ConfigValidators(ctx) {
+			vpcRes := &ValidateProviderConfigResponse{}
+
+			configValidator.Validate(ctx, vpcReq, vpcRes)
+
+			if vpcRes.Diagnostics != nil {
+				resp.Diagnostics = append(resp.Diagnostics, vpcRes.Diagnostics...)
+			}
+		}
+	}
+
+	if provider, ok := s.p.(ProviderWithValidateConfig); ok {
+		vpcRes := &ValidateProviderConfigResponse{}
+
+		provider.ValidateConfig(ctx, vpcReq, vpcRes)
+
+		if vpcRes.Diagnostics != nil {
```

Comment:
Is it worth using `if len(vpcRes.Diagnostics) > 0`? I don't know that this check is necessary at all, honestly, but.

Reply:
It is not 😄  I will remove the conditional.

Reply:
Removed these across the board in 209de761804e2367f0fd0e9cdcc513146cc1bae5 👍